arrows/core/render_mesh_depth_map.cxx

```diff
@@ -0,0 +1,203 @@
+  {
+    points_2d[i] = camera->project(vertices[i]);
+  }
+
+  // Compute the points negative inverse depth
+  std::vector<double> points_depth(nb_vertices);
+  for (unsigned int i=0; i < vertices.size(); ++i)
+  {
+    points_depth[i] = -1.0 / camera->depth(vertices[i]);
+  }
+
+  // Initialize z_buffer with max infinity
+  vital::image_of<double> z_buffer(camera->image_width(), camera->image_height(), 1);
+  for (size_t j = 0; j < z_buffer.height(); ++j)
+  {
+    for (size_t i = 0; i < z_buffer.width(); ++i)
+    {
+      z_buffer(i, j) = std::numeric_limits<double>::infinity();
+    }
+  }
+
+  rasterize_depth(mesh, points_2d, points_depth, z_buffer);
+
+  // Re-compute the real depth
+  for (size_t j = 0; j < z_buffer.height(); ++j)
+  {
+    for (size_t i = 0; i < z_buffer.width(); ++i)
+    {
+      if (!std::isinf(z_buffer(i, j)))
+      {
+        z_buffer(i, j) = -1.0 / z_buffer(i, j);
+      }
+    }
+  }
+  return std::make_shared<vital::simple_image_container>(z_buffer);
+}
+
+
+void rasterize_depth(vital::mesh_sptr mesh, const std::vector<vital::vector_2d> &image_points,
+                     const std::vector<double> &depth, vital::image_of<double>& image)
+{
+  int width = image.width();
+  int height = image.height();
+  // Write faces on z_buffer with depth test
+  vital::mesh_face_array& faces = dynamic_cast< vital::mesh_face_array& >(mesh->faces());
+  for (unsigned int f_id = 0; f_id < faces.size(); ++f_id)
+  {
+    const vital::vector_2d& a_uv = image_points[faces(f_id, 0)];
+    const vital::vector_2d& b_uv = image_points[faces(f_id, 1)];
+    const vital::vector_2d& c_uv = image_points[faces(f_id, 2)];
+
+    // skip the face if the three points are outside the image
+    if ((a_uv[0] < 0 || a_uv[0] >= width || a_uv[1] < 0 || a_uv[1] >= height) &&
+        (b_uv[0] < 0 || b_uv[0] >= width || b_uv[1] < 0 || b_uv[1] >= height) &&
+        (c_uv[0] < 0 || c_uv[0] >= width || c_uv[1] < 0 || c_uv[1] >= height))
+      continue;
+    double a_depth = depth[faces(f_id, 0)];
+    double b_depth = depth[faces(f_id, 1)];
+    double c_depth = depth[faces(f_id, 2)];
+
+    // rasterization is done over the face axis-aligned bounding box
+    int u_min = std::max(0, static_cast<int>(std::round(std::min(a_uv[0], std::min(b_uv[0], c_uv[0])))));
+    int u_max = std::min(width - 1, static_cast<int>(std::round(std::max(a_uv[0], std::max(b_uv[0], c_uv[0])))));
+    int v_min = std::max(0, static_cast<int>(std::round(std::min(a_uv[1], std::min(b_uv[1], c_uv[1])))));
+    int v_max = std::min(height - 1, static_cast<int>(std::round(std::max(a_uv[1], std::max(b_uv[1], c_uv[1])))));
+
+    for (int v = v_min; v <= v_max; ++v)
+    {
+      for (int u = u_min; u <= u_max; ++u)
+      {
+        vital::vector_2d p(u, v);
+
+        // Handle pixels on triangle boundaries. Assignment rules:
+        //  - if the pixel center is inside the triangle
+        //  - if the pixel is not alread assigned and if the pixel intersects the triangle
```

Comment:
What is the motivation for this special case of point near the boundary if the pixel is not already assigned?  This could give different rendering results depending on the order in which you process triangles.

also, typo in comment: "alread" --> "already"

Reply:
The motivation is to include pixels which partially intersect with the triangle. Without this, the pixels where the triangle corners are located might be empty.

The result depends on the order in which I process the triangles in some cases, but it is quite rare (the pixel is intersected by more than one triangle and the center of this pixel is outside each of these triangles).